File: app/src/main/java/com/greglaun/lector/TtsPresenter.java
package com.greglaun.lector;

import android.speech.tts.UtteranceProgressListener;

import com.greglaun.lector.ui.speak.TTSContract;
import com.greglaun.lector.data.model.speakable.TmpTxtBuffer;

import org.jetbrains.annotations.NotNull;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class TtsPresenter extends UtteranceProgressListener
        implements TTSContract.Presenter {
    public interface Callback { void call(); }

    private TextProvider provider;
    private volatile boolean speaking;
    private final TTSContract.AudioView tts;
    TmpTxtBuffer buffer = new TmpTxtBuffer();
    Executor speechExecutor = Executors.newSingleThreadExecutor();
    private Callback endOfArticleCallback;

    public TtsPresenter(TTSContract.AudioView tts) {
        this.tts = tts;
    }

    @Override
    public void onUrlChanged(@NotNull String urlString) {
        // todo: implement
    }

    private void queueForSpeaking(String text) {
        tts.speak(text);
    }

    public void startSpeaking(final TextProvider provider) {
        this.provider = provider;
        buffer.addFromProvider(provider);
        speaking = true;
        speechExecutor.execute(new MainSpeechLoop());
    }

    @Override
    public void startSpeaking() {
        // todo: Implement method
    }

    @Override
    public void stopSpeaking() {
        // todo: Implement method
    }

    public TextProvider getProvider() {
        return provider;
    }

    public void setProvider(TextProvider provider) {
        this.provider = provider;
    }

    private class MainSpeechLoop implements Runnable {
        @Override
        public void run() {
            playOneIfSpeaking();
        }
    }

    public void playOneIfSpeaking() {
        if (speaking) {
            String textToSpeak = buffer.getCurrent();
            if (textToSpeak.equals(TextProvider.END_OF_STREAM)) {
                if (endOfArticleCallback != null) {
                    endOfArticleCallback.call();
                }
                return;
            }
            queueForSpeaking(textToSpeak);
        }
    }

    public void flush() {
        buffer.clear();
    }

    public String getCurrentUtterance() {
        return buffer.getCurrent();
    }

    public void setEndOfArticleCallback(Callback endOfArticleCallback) {
        this.endOfArticleCallback = endOfArticleCallback;
    }

    @Override
    public void onStart(String utteranceId) {

    }

    @Override
    public void onDone(String utteranceId) {
        buffer.advance();
        playOneIfSpeaking();
    }

    @Override
    public void onError(String utteranceId) {

    }

//    private fun isWikiUrl(url: String): Boolean {
//        return (url.startsWith("https://" + WIKI_LANGUAGE + ".wikipedia.org/wiki")
//                || url.startsWith(
//                "https://" + WIKI_LANGUAGE + ".m.wikipedia.org/wiki")
//                && !url.contains("File:")
//        )
//    }


}

File: app/src/main/java/com/greglaun/lector/TtsPresenter.kt
package com.greglaun.lector

import android.net.Uri
import android.util.Log
import com.greglaun.lector.data.cache.utteranceId
import com.greglaun.lector.data.model.speakable.TmpTxtBuffer
import com.greglaun.lector.ui.main.MainContract
import com.greglaun.lector.ui.speak.TTSContract
import kotlinx.coroutines.experimental.GlobalScope
import kotlinx.coroutines.experimental.Job
import kotlinx.coroutines.experimental.launch
import java.util.concurrent.Semaphore

class TtsPresenter(private val tts: TTSContract.AudioView,
                   private val mainPresenter : MainContract.Presenter) : TTSContract.Presenter {
    private val TAG = TtsPresenter::class.java.simpleName

    val readyLock  = Semaphore(1)
    var articleStarted = false
    var provider: TextProvider? = null
    private var buffer = TmpTxtBuffer()
    private var mainLoop : Job? = null

    override fun onUrlChanged(urlString: String) {
        mainPresenter.stopSpeakingAndEnablePlayButton()
        GlobalScope.launch {
            try {
                provider = JSoupTextProvider.createFromUri(Uri.parse(urlString)).await()
                if (provider != null) {
                    // todo(mvp): Remove dependency on JoupTextProvider
                    buffer.clear()
                    buffer.addFromProvider(provider as JSoupTextProvider)
                }
                articleStarted = true
            } catch (e: Exception) {
                Log.d(TAG, "JSoupTextProvider failed", e)
            }
        }
    }


    override fun startSpeaking() {
        // todo(concurrency): Is a semaphore really the right solution here?
        mainLoop = GlobalScope.launch {
           if (buffer?.isExhausted() && articleStarted) {
                mainPresenter.onArticleOver()
                return@launch
            }
            readyLock.release()
            while (true) {
                if (buffer.isExhausted()) {
                    mainPresenter.onArticleOver()
                }
                readyLock.acquire()
                val text = buffer.getCurrent()
                tts.speak(text) {
                    if (it == utteranceId(text)) {
                        buffer.advance()
                        readyLock.release()
                    }

                }
            }
        }
    }

    override fun stopSpeaking() {
        mainLoop?.cancel()
        tts.stopImmediately()
    }

    //    private fun isWikiUrl(url: String): Boolean {
    //        return (url.startsWith("https://" + WIKI_LANGUAGE + ".wikipedia.org/wiki")
    //                || url.startsWith(
    //                "https://" + WIKI_LANGUAGE + ".m.wikipedia.org/wiki")
    //                && !url.contains("File:")
    //        )
    //    }


}
